### ISSUE TITLE
allow forbid_action to accept a list? :edit, :create, :show etc

### ISSUE THREAD

**Issue by jpwynn** [chrisalley/pundit-matchers#8]
First, this gem is awesome for succinctly handling robust policy testing specs.

It seems it would be useful (and relatively easy) if forbid_action and permit_action would accept more than one action, eg: if instead of:

    context "foobar" do
          it { is_expected.to permit_action(:aaa)    }
          it { is_expected.to forbid_action(:bbb)  }
          it { is_expected.to permit_action(:ccc)  }
          it { is_expected.to forbid_action(:ddd)     }
          it { is_expected.to permit_action(:eee)  }
          it { is_expected.to forbid_action(:fff)    }
          it { is_expected.to permit_action(:ggg)    }
    end

a spec could say:

    context "foobar" do
          it { is_expected.to permit_action( :aaa, :ccc, :eee, :ggg )    }
          it { is_expected.to forbid_action( :bbb, :ddd, :fff )  }
    end

To me, that is much easier to scan when reviewing/sanity checking dozens of policy scenarios

Furthermore, as a freebie, it would nicely lend itself to yet another useful level of abstraction, the specification of common 'sets' of permitted/forbidden actions, eg:

    context_foobar = { :permitted => [:aaa, :ccc, :eee, :ggg], :forbidden => [:bbb, :ddd, :fff]}
    context_barfoo = { :permitted => [:aaa, :eee, ], :forbidden => [:bbb, :ccc, :ddd, :fff, :ggg]}

which then automatically allows collapsing this:

    context "foobar" do
          it { is_expected.to permit_action(:aaa)    }
          it { is_expected.to forbid_action(:bbb)  }
          it { is_expected.to permit_action(:ccc)  }
          it { is_expected.to forbid_action(:ddd)     }
          it { is_expected.to permit_action(:eee)  }
          it { is_expected.to forbid_action(:fff)    }
          it { is_expected.to permit_action(:ggg)    }
    end
    context "barfoo" do
          it { is_expected.to permit_action(:aaa)    }
          it { is_expected.to forbid_action(:bbb)  }
          it { is_expected.to forbid_action(:ccc)  }
          it { is_expected.to forbid_action(:ddd)     }
          it { is_expected.to permit_action(:eee)  }
          it { is_expected.to forbid_action(:fff)    }
          it { is_expected.to forbid_action(:ggg)    }
    end

to this:

          it { is_expected.to permit_action(context_foobar[:permitted])    }
          it { is_expected.to forbid_action(context_foobar[:forbidden])    }
          it { is_expected.to permit_action(context_barfoo[:permitted])    }
          it { is_expected.to forbid_action(context_barfoo[:forbidden])    }

Finally, for anyone else who tests all controller actions, when adding a new action you just have to add it to each of the context hashes that list the actions, instead adding a new permit/forbid line to each of many blocks.

Currently, if you supply a list of actions, it seems to use the first and ignore the rest, which provides somewhat misleading results since it reports as of all where tested:

          (passed) should forbid action :show, :publish, :index, :create, :new, :update, :edit, and :destroy

when, in this case, there is not even a authorize statement for the newly added publish action, which should have triggered the verify_authorized exception

**Comment by chrisalley** [comment]
To keep the DSL close to the English language, I think that we should add new matchers called `permit_actions` and `forbid_actions` that can be used for multiple actions, while reserving `permit_action` and `forbid_action` for specifying whether a single action is permitted or not. E.g.

```ruby
it { is_expected.to permit_action(:publish) }
```

```ruby
it { is_expected.to permit_actions(:publish, :unpublish) }
```

Potentially we could make `permit_actions` flexible enough so that if an array is passed to it, it would check the entire of contents of the array. This would allow a list of actions to be reused across multiple assertions. E.g.

```ruby 
administrative_actions = [:publish, :unpublish]

context 'being a visitor' do
  let(:user) { nil }
  it { is_expected.to forbid_actions(administrative_actions) }
end
  
context 'being an administrator' do
  let(:user) { User.create(administrator: true) }
  it { is_expected.to permit_actions(administrative_actions) }
end
```

I've created a [seperate issue](https://github.com/chrisalley/pundit-matchers/issues/9) for `permit_action` accepting multiple actions, which was not the intended behaviour.

**Comment by jpwynn** [comment]
FWIW, here is a quick and dirty version I use now in a file I add to my spec/support folder. What I did not figure out is how to have them work correctly in the .to_not context, but for me that is not important since there is both a permit_actions and forbid_actions so not sure where a .to_not would be useful.

My policy specs are now very tight, and IMO easier to read and understand, since there are now typically 3 or 4 recurring 'named patterns' of permit/forbid based on the user role that get defined once up at the top of the spec, for example:

    # define the users and records for various scenarios 
    # like active/inactive author and or admins and/or public users, 
    #  acting upon published/unpublished posts
    let(:author_active_agent) { FactoryGirl.create(:agent, :with_confirmed, :set_active) }
    let(:published_post_by_active_author) { FactoryGirl.create(:post, status: :published, agent: author_active_agent) }
    let(:unpublished_post_by_active_author) { FactoryGirl.create(:post, status: :unpublished, agent: author_active_agent) }
    .... more actors and records

    # define each common permit/forbid patterns as hash

    # public users, published posts for example
    let(  :context_show_only  ) { { 
        :permitted => [:show], 
        :forbidden => [:index, :create, :new, :update, :edit, :destroy, :publish, :unpublish] 
      }}

    # public users unpublished posts, for exmaple
    let(  :context_show_only  ) { { 
        :permitted => [], 
        :forbidden => [:show, :index, :create, :new, :update, :edit, :destroy, :publish, :unpublish] 
      }}

    let(  :context_agents_own_published_record  ) { { 
        :permitted => [:show,  :update, :edit,  :unpublish], 
        :forbidden => [:index,  :destroy, :publish ] 
      }}

    let(  :context_agents_own_unpublished_record  ) { { 
        :permitted => [:show,  :update, :edit, :publish], 
        :forbidden => [:index,  :destroy, :unpublish ] 
      }}

    # admins can do anything
    let(  :context_everything_allowed  ) { { 
        :permitted => [:show, :index, :create, :new, :update, :edit, :destroy, :publish, :unpublish], 
        :forbidden => [] 
      }}

so in the spec for an agent acting on its own record:

      context 'for active agent acting upon its own'  do
        let(:user) { author_active_agent }
        context "published post" do
          let(:record) { published_post_by_active_author }

          it { is_expected.to forbid_actions( context_agents_own_published_record[:forbidden] )  }
          it { is_expected.to permit_actions( context_agents_own_published_record[:permitted] )  }

        end

        context "unpublished post" do
          let(:record) { unpublished_post_by_active_author }

          it { is_expected.to forbid_actions( context_agents_own_unpublished_record[:forbidden] )  }
          it { is_expected.to permit_actions( context_agents_own_unpublished_record[:permitted] )  }

        end

The quick & dirty matchers I added are below... note the specific failing action is reported:

```
RSpec::Matchers.define :permit_actions do |actions|
  failed = []
  match do |policy|
    r = true
    actions.each do |action|
      isok = policy.public_send("#{action}?")
      failed << action if !isok
      r = r && isok
    end
    return r
  end

  failure_message do |policy|
    "#{policy.class} does not permit #{failed} on #{policy.record} for " \
      "#{policy.user.inspect}."
  end

  failure_message_when_negated do |policy|
    "DONT USE NEGATED FORM OF THSI MATCHER #{policy.class} does not forbid #{failed} on #{policy.record} for " \
      "#{policy.user.inspect}."
  end
end

RSpec::Matchers.define :forbid_actions do |actions|
  failed = []
  match do |policy|
    r = true
    # r = false if actions.empty?
    actions.each do |action|
      isok = !policy.public_send("#{action}?")
      failed << action if !isok
      r = r && isok
    end
    return r
  end

  failure_message do |policy|
    "#{policy.class} does not forbid #{failed} on #{policy.record} for " \
      "#{policy.user.inspect}."
  end

  failure_message_when_negated do |policy|
    "DONT USE NEGATED FORM OF THSI MATCHER #{policy.class} does not forbid #{actions} on #{policy.record} for " \
      "#{policy.user.inspect}."
  end
end
```

**Comment by chrisalley** [comment]
With the negative exceptions, we could check that each action is forbidden or permitted, e.g.

```ruby
  it { is_expected.to_not forbid_actions(administrative_actions) }
```

...would check that each administrative action in the array is permitted and fail if one or more action is not permitted.

**Comment by chrisalley** [comment]
This is now implemented (in pundit-matchers 1.2.1).
